qpext/cmd/qpext/main.go

```diff
@@ -80,24 +81,109 @@ func applyHeaders(into http.Header, from http.Header, keys ...string) {
 	}
 }
 
-func negotiateMetricsFormat(contentType string) expfmt.Format {
-	mediaType, _, err := mime.ParseMediaType(contentType)
-	if err == nil && mediaType == expfmt.OpenMetricsType {
-		return expfmt.FmtOpenMetrics
+func getServerlessLabelVals() []string {
+	var labelValues []string
+	for _, envVar := range EnvVars {
+		labelValues = append(labelValues, os.Getenv(envVar))
 	}
-	return expfmt.FmtText
+	return labelValues
 }
 
-// scrapeAndWriteAgentMetrics gathers a slice of prometheus metric families and encodes the metrics.
-func scrapeAndWriteAgentMetrics(w io.Writer) error {
-	mfs, err := promRegistry.Gather()
-	if err != nil {
-		return err
+// addServerlessLabels adds the serverless labels to the prometheus metrics that are imported in from the application.
+// this is done so that the prometheus metrics (both queue-proxy's and kserve-container's) can be easily queried together.
+func addServerlessLabels(metric *io_prometheus_client.Metric, labelKeys []string, labelValues []string) *io_prometheus_client.Metric {
+	// LabelKeys, EnvVars, and LabelVals are []string to enforce setting them in order (helps with testing)
+	for idx, name := range labelKeys {
+		labelName := name
+		labelValue := labelValues[idx]
+		newLabelPair := &io_prometheus_client.LabelPair{
+			Name:  &labelName,
+			Value: &labelValue,
+		}
+		metric.Label = append(metric.Label, newLabelPair)
+	}
+	return metric
+}
+
+// sanitizeMetrics attempts to convert UNTYPED metrics into either a gauge or counter.
+// counter metric names with _created and gauge metric names with _total are converted due to irregularities
+// observed in the conversion of these metrics from text to metric families.
+func sanitizeMetrics(mf *io_prometheus_client.MetricFamily) *io_prometheus_client.MetricFamily {
+	if strings.HasSuffix(*mf.Name, "_created") {
+		counter := io_prometheus_client.MetricType_COUNTER
+		var newMetric []*io_prometheus_client.Metric
+		for _, metric := range mf.Metric {
+			newMetric = append(newMetric, &io_prometheus_client.Metric{
+				Label: metric.Label,
+				Counter: &io_prometheus_client.Counter{
+					Value: metric.Untyped.Value,
+				},
+				TimestampMs: metric.TimestampMs,
+			})
+		}
```

Comment:
This block and the block from line 134-143 can be extracted out of the if condition and shared between the two if checks.

Reply:
the types are different one is counter one is gauge

Reply:
but I can make a helper method out of it Ithink

Reply:
discussed - probably not saving much code/readability with a helper method here